### PR TITLE
PCG follow up changes

### PR DESCRIPTION
**Pull request by mppf**
This is a follow-on to PR #3506.

* Move Random.SeedGenerator instance to Random.SeedGenerator type with same methods as type methods
* Split low-level PCG functions into PCGRandomLib module
* Move NBPRandom module last in file
* Updates tests to replace SeedGenerator.currentTime with SeedGenerator.oddCurrentTime (odd in case the test is using NPB)
* Reordered code in Random.chpl to be in an order that makes more sense when reading the documentation.
* Address much of Brad's documentation feedback
* Fixes a bug that occurred to me do with resultType support and improves test to catch the problem

Reviewed by @bradcray - thanks!

Still to-do (in a separate PR):

 * look at broken links

Passed full local testing.

### CODE REVIEW

**Comment by mppf** [comment]
@bradcray - could you review please? It is a follow on to the PCG PR you already reviewed and addresses some of your feedback. Also, do you have any second thoughts about deprecating SeedGenerator, given the number of tests I needed to change? (I don't personally have a problem with leaving the old one there for a while without a warning... )

**Comment by bradcray** [comment]
This part about SeedGenerator caught my eye on email:  If we're going to remove SeedGenerator in the next release (which is what I assumed deprecating it meant), won't this number of tests have to change in the future at that point if not now?  Or did you imagine a different path for deprecating it?

**Comment by mppf** [comment]
Right. The alternative is not deprecating it now - not removing it in 1.14. It's small and somewhat trivial. Given that we're expecting other design-level reworking in Random at some point, we could wait for that effort to deprecate it. I don't have a strong opinion on this but changing those tests for such a little thing seemed much ado about nothing.

**Comment by bradcray** [comment]
If I'd been implementing this, I probably would've just renamed the 'SeedGenerators' record in 1.12 into 'SeedGenerator', dropped the 'SeedGenerator' variable and made 'currentTime' into a type method on it, figuring that the number of people who cared about how SeedGenerator was defined and what the name of its type was is probably near-zero.

But in yesterday's PR, it looked like you were going to some trouble to preserve the old type and variable, which I assumed was to help people transition into the new names, which I thought was admirable.  But assuming that was the intention, it seemed like we should add a warning to them indicating that it was a temporary measure while they switched over to the new type methods, which was what my comment on yesterday's PR was trying to suggest.

We could back off from taking the high road and go back to the plan I would've done which would negate the need to change 'SeedGenerator' to 'SeedGenerators' in the tests (which I think is the preferable name anyway -- the plural seems grammatically slightly worse in the usage context -- the type itself is a "seed generator" rather than a collection of "seed generators").  But even in this approach, it seems we'd need to switch the method called to request the odd time for the NPB stream?  At least, I've lost track of the details of how we got away with it previously but require it now (though I strongly suspect that your approach is more on the up-and-up).  So I don't think doing this would reduce the number of test changes...

**Comment by mppf** [comment]
Right. What if we renamed 'SeedGenerators' into 'SeedGenerator' ? Then we would have a problem because currentTime exists in both and does something different. I suppose that is enough to justify the deprecation, since there *is* a behavior change. The symptom of the problem would be that programs that used to work and were manually switched to NPB would fail with a halt if the current time was even.

That seems somewhat unlikely to me though. (Outside of the great number of tests we have, I'd expect programs to work when the RNG switches to PCG without modification - and so I don't expect Chapel programmers to manually request NPB for their programs).

Another possible path would be to just make currentTime always return an odd time, even though that might not be necessary for some RNGs. It seems a slippery slope, but would reduce some of this problem and may be reasonable in practice.

Anyway, if we keep it the way this PR does it - I think it would be worthwhile to add something to the deprecation warning to indicate that currentTime should be changed to oddCurrentTime.

So, summing up, I think we have 3 choices:
 1) SeedGenerators -> SeedGenerator, currentTime -> oddCurrentTime  (what this PR is now)
 2) SeedGenerator keeps name. currentTime -> oddCurrentTime but only necessary for NPB
 3) SeedGenerator keeps name. currentTime produces only odd values.

At the moment, I'm thinking that 2 is perhaps the best approach, but I'd appreciate your view on it.

Re. how we got away without oddCurrentTime before now - my initial PCG PR made the NPB seeding function adjust the provided seed to be odd, but you convinced me to remove that change.

File: modules/standard/Random.chpl [hunk 1]
```diff
@@ -601,15 +841,15 @@ module Random {
       which has 64 bits of state and 32 bits of output.
 
       While the PCG RNG used here is believed to have good statistical
-      properties, it is not suitable for cryptographic use. Additionally,
-      if statistical properties of the random numbers are very important,
-      another strategy may be required.
+      properties, it is not suitable for generating key material for encryption.
```
Comment (bradcray) [inline modules/standard/Random.chpl]: Given the explanation you gave in the PR conversation, I think this would be better if it gave the rationale (good cryptologists will know presumably, but so that the comment makes sense to the rest of us):  E.g., I'm imagining tacking on something like "due to the predictability (reproducibility?  determinism?) of the values generated by the stream."

File: modules/standard/Random.chpl [hunk 2]
```diff
@@ -80,7 +80,12 @@ module Random {
 
     // Will be deprecated in 1.14.
     pragma "no doc"
-    const SeedGenerator: SeedGeneratorsDeprecated;
+    const SeedGenerator_actual: SeedGeneratorsDeprecated;
+    pragma "no doc"
+    proc SeedGenerator {
+      compilerWarning("Random.SeedGenerator is deprecated - use Random.SeedGenerators");
+      return SeedGenerator_actual;
+    }
```
Comment (bradcray) [inline modules/standard/Random.chpl]: I don't think this is worth changing at this point since it's code that's going to go away anyway, but this compilerWarning() could've potentially been put into a constructor for SeedGeneratorDeprecated as well to avoid the need for the proc + _actual variable.  (Or into its currentTime query on the argument that any meaningful use of SeedGenerator would call the routine?  even though that's not as bulletproof...)

File: modules/standard/Random.chpl [hunk 3]
```diff
@@ -1053,6 +1054,7 @@ module Random {
     }
 
 
+    module PCGRandomLib {
```
Comment (bradcray) [inline modules/standard/Random.chpl]: Ooh, is this our first sub-sub-module?  Sweeeet!  Can't wait to see it under Lydia's new docs rendering...!

File: modules/standard/Random.chpl [hunk 4]
```diff
@@ -566,6 +574,10 @@ module Random {
      imag, and complex numbers; and generating numbers in a range in parallel.
      These features are not available in the reference implementations of PCG.
 
+     .. note::
+
+       This interface provided by this module is expected to change.
```
Comment (bradcray) [inline modules/standard/Random.chpl]: Really minor but "The interface" scans better for me... not worth it unless you've got other changes to make, though

File: modules/standard/Random.chpl [hunk 5]
```diff
@@ -590,15 +602,15 @@ module Random {
...
-      We have verified that the random numbers generated by this class can match
-      those generated by the C PCG reference implementation. However, this
-      implementation differs from the C PCG reference implementation in how it
-      produces random integers in a range and in support for generating random
-      `real` s.
+      We have good confidence that the random numbers generated by this class
+      match the C PCG reference implementation and have specifically verified
+      equal output given the same seed. However, this implementation differs
+      from the C PCG reference implementation in how it produces random integers
+      in a range and in support for generating random `real` s.
```
Comment (bradcray) [inline modules/standard/Random.chpl]: This is a nice change overall.  Naming the routine used to generate bounded integers might help clarify this point.  Specifically "random integers in a range" is easily confused with Chapel's notion of "range"

File: modules/standard/Random.chpl [hunk 6]
```diff
@@ -661,9 +673,8 @@ module Random {
     class RandomStream {
       /*
         Specifies the type of value generated by the PCGRandomStream.
-        Currently, only numeric types are supported, but all numeric
-        types should be supported (int, uint, of all sizes,
-        real, imag, and complex types of all sizes, and bool types).
+        All numeric types are supported: int, uint, of all sizes,
+        real, imag, and complex types of all sizes, and bool types.
```
Comment (bradcray) [inline modules/standard/Random.chpl]: Probably bool types of all sizes are supported too.  :)  So this could simply say "bool, int, uint, real, imag, and complex types of all sizes."  I don't know if bools are really numeric types...

(Man, if only we added support for random voids and strings, we could claim all the primitive types are supported... :)  ).

File: modules/standard/Random.chpl [hunk 7]
```diff
@@ -775,9 +788,10 @@ module Random {
 
       /*
         Advances/rewinds the stream to the `n`-th value in the sequence.
-        The first value is with n=1.
+        The first corresponds to n=1. It is an error to call this
+        routine with n <= 0.
 
-        :arg n: The position in the stream to skip to.  Must be non-negative.
+        :arg n: The position in the stream to skip to.  Must be > 0.
```
Comment (bradcray) [inline modules/standard/Random.chpl]: I was thinking that NPB was different in this regard, but it turns out its documentation is just wrong too.  Could you fix its documentation as well?  (sorry, I very well was the person who introduced this mistake).

File: modules/standard/Random.chpl [hunk 8]
```diff
@@ -2114,9 +2128,14 @@ module Random {
     :mod:`PCGRandom` and :mod:`NPBRandom` for RNGs that can be
     instantiated. To create a random stream, use :proc:`makeRandomStream`.
 
-    Note that whether or not 0.0 and 1.0 are possiblly generated real
-    numbers varies. (The can be generated by :mod:`PCGRandom`
-    but not by :mod:`NPBRandom`).
+    Note different implementations of this interface can vary in whether or not
+    they can generate 0.0 and 1.0.  (e.g. The can be generated by
```
Comment (bradcray) [inline modules/standard/Random.chpl]: "The" -> "They".  While here, might change "and" into "and/or" (since we might conceive of future RNGs that generate one but not the other?)

File: modules/standard/Random.chpl (hunk 3 above)
Comment (mppf) [inline modules/standard/Random.chpl]: The later commit moves it out to a sister module to PCGRandom, which I think is better code organization anyway.

File: modules/standard/Random.chpl [hunk 9]
```diff
@@ -549,30 +809,19 @@ module Random {
...
 
+     .. note::
+
+       This interface provided by this module is expected to change.
+
   */
   module PCGRandom {
 
     use RandomSupport;
+    use PCGRandomLib;
```
Comment (bradcray) [inline modules/standard/Random.chpl]: This is the type of use that I think we'd want to be a 'private use' (non-transitive) if/when we get that concept.  In reviewing code, I usually encourage people to push use's like this into the scopes where the PCGRandomLib symbols are used directly (or to use the fully-qualified names to refer to such symbols) in order to avoid bleeding symbols out of the module unintentionally, but given how widespread this probably is in this routine and the timeframe, maybe that's not reasonable.

File: modules/standard/Random.chpl [hunk 10]
```diff
@@ -1068,6 +1052,27 @@ module Random {
     }
 
 
+    /*
+
+     Low-level PCG RNG implementation.
+
+     Besides a Chapel :class:`RandomStream` interface, the :mod:`PCGRandomLib`
```
Comment (bradcray) [inline modules/standard/Random.chpl]: w.r.t. this introductory clause, I may not be tracking properly, but this sub-submodule doesn't still supply a RandomStream interface does it?

In retrospect, I wonder if there's a better name for this module.  Namely, if you asked me out of the blue "What's the difference between PCGRandom and PCGRandomLib, I'm not sure I have any intuition.  The latter is what Kyle was considering the "system" interface, but I'm not wild about that label/distinction either.  PCGRandomLowLevel is the first thing that comes to mind for me, though that's somewhat of a mouthful...  or PCGRandomUtil?  PCGRandomHelp?

File: modules/standard/Random.chpl [hunk 11]
```diff
@@ -1208,143 +1208,425 @@ module Random {
...
+        return (rand32_1(states) >> 31) != 0;
+      }
+    }
 
-     */
-    module PCGRandomLib {
-    /*
-       Low-level PCG random number generation interface (64-bits of state,
-       32-bits output).
+    // returns x with min <= x <= max (for integers)
```
Comment (bradcray) [inline modules/standard/Random.chpl]: I can't make sense of this diff, so am relying on it simply being a code reordering and not reviewing further.  I could imagine going further and putting PCG ahead of NPB in the file on the argument that it's the dominant case for those reading the source.

**Comment by bradcray** [comment]
> Then we would have a problem because currentTime exists in both and does something different.

But wouldn't one be a type method and the other a value method such that they could do something different?  (I easily could still be missing something else).

**Comment by bradcray** [comment]
(BTW, I think I'm done reviewing.  I don't have a problem with the current handling of the deprecation of seed generator per se, but am happy to keep discussing alternatives if you'd like to keep wrestling with it... that last comment being my current sticking point).

**Comment by mppf** [comment]
> > Then we would have a problem because currentTime exists in both and does something different.

> But wouldn't one be a type method and the other a value method such that they could do something  different?  (I easily could still be missing something else).

I think we both doubt that anybody actually stored Random.SeedGenerator into a variable (e.g.)

      var x = SeedGenerator

instead, the code we see uses a pattern like

     var x = if randomSeed then SeedGenerator.currentTime else 41;

For this pattern, it does not matter if SeedGenerator is a record with a regular method called currentTime, or if SeedGenerator is a type with a type method currentTime. If that was all that needed to change, the old code would work. Since this seems to be the common way to use SeedGenerator at all, it seems appealing to not require a code change in this case. I'm still not sure this is a big enough deal for a deprecation warning (although it did help me find the tests I needed to change to give NPB an odd seed).

**Comment by bradcray** [comment]
Oh, I see the mistake in my thinking.  Rats.

OK, I think I agree that of your choices, #2 is most attractive.  If I'm understanding correctly, the outcome would be:

* "I use RandomStream and SeedGenerator, but don't really care whether I get NPB or PCG and would prefer whatever's best (PCG)" == no changes

* "I'm a test that uses RandomStream and requires NPB (because my output is sensitive to it; because I'm a NPB benchmark and we want to do apples to apples comparisons)" == requires changing from SeedGenerator.currentTime to SeedGenerator.oddCurrentTime

And we get no warning about SeedGenerators being deprecated because it isn't compatible with the above and we don't think it will matter to anyone sufficiently anyway.

Is that right?  If so, I agree this is the way to proceed.

File: modules/standard/Random.chpl (hunk 10 above)
Comment (mppf) [inline modules/standard/Random.chpl]: I think I'm going to leave it as PCGRandomLib as PCGRandomUtil/Help/LowLevel seem worse to me.
I think we're going to need a strategy for naming such things and that we should change PCGRandomLib to use that strategy once we know what it is.
